### PR TITLE
✨(api) add users endpoint for deletion management

### PR DESCRIPTION
**Pull request by wilbrdt**
## Purpose

As discussed in https://github.com/openfun/mork/issues/44, Mork should be the central API that exposes endpoints so that each application can retrieve the list of users it needs to delete. 

## Proposal

_Note: We still need to populate the new tables at deletion time from the edX MySQL database.
It will be done in a next PR._

### List Users
- **Endpoint:** `GET /v1/users`
- **Description:** Retrieves a list of users based on the query parameters
- **Query Parameters:**
  - `deletion_status` (optional): Filter users by deletion status
  - `service` (optional): The name of the service to filter users on
  - `limit` (optional): Maximum number of items to retrieve (default: 100, max: 1000)
  - `offset` (optional): Number of items to skip (default: 0)
- **Response:** `200 OK`
```json
[
  {
    "id": "id123",
    "username": "user123",
    "email": "user123@example.com",
    ...
  },
  {
    "id": "id456",
    "username": "user456",
    "email": "user456@example.com",
    ...
  }
]
```

### Read Specific User
- **Endpoint:** `GET /v1/users/{user_id}`
- **Description:** Retrieve a user by their ID
- **Path Parameters:**
  - `user_id`: The ID of the user to read
- **Query Parameters:**
  - `service` (optional): The name of the service to filter on
- **Response:** `200 OK`
```json
{
  "id": "id123",
  "username": "user123",
  "email": "user123@example.com",
  ...
}
```
- **Error Response:** `404 Not Found` if user doesn't exist

### Read User Status
- **Endpoint:** `GET /v1//users/{user_id}/status/{service_name}`
- **Description:** Read the user deletion status for a specific service
- **Path Parameters:**
  - `user_id`: The ID of the user
  - `service_name`: The name of the service
- **Response:** `200 OK`
```json
{
  "id": "id123",
  "service_name": "service_name",
  "status": "deletion_status"
}
```
- **Error Response:** `404 Not Found` if status doesn't exist

### Update User Status
- **Endpoint:** `PATCH /v1//users/{user_id}/status/{service_name}`
- **Description:** Update the user deletion status for a specific service
- **Path Parameters:**
  - `user_id`: The ID of the user
  - `service_name`: The name of the service
- **Request Body:**
```json
{
  "deletion_status": "deleted"
}
```
- **Response:** `200 OK`
```json
{
  "id": "id123",
  "service_name": "service_name",
  "status": "deleted"
}
```
- **Error Response:** `404 Not Found` if status doesn't exist

### CODE REVIEW

File: src/app/mork/models/users.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+    USER_REQUESTED = "user_requested"
+    GDPR = "gdpr"
+
+
+class DeletionStatus(str, enum.Enum):
+    """Enum for deletion statuses."""
+
+    TO_DELETE = "to_delete"
+    DELETED = "deleted"
```
Comment (lunika) [inline src/app/mork/models/users.py:41]: IMO a `deleting` status should be useful.

We should have `to_delete` -> `deleting` -> `deleted`. Doing this, we can prevent race condition by preventing to process twice the same user

File: src/app/mork/api/v1/routers/users.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+from mork.schemas.users import DeletionStatus, UserRead, UserUpdate
+
+logger = logging.getLogger(__name__)
+
+router = APIRouter(prefix="/users", dependencies=[Depends(authenticate_api_key)])
+
+
+@router.get("")
+@router.get("/")
+async def read_users(
```
Comment (lunika) [inline src/app/mork/api/v1/routers/users.py:27]: IMO adding a  `read_user` allowing to fetch a specific user can be useful. Doing this, we can have the state of a specific user at a specific time. 
We can also test at the beginning of each deletion process in each application if the user is still to process or not

Comment (wilbrdt) [reply src/app/mork/api/v1/routers/users.py:27]: I can add a `read_user` as it sure will be useful.
Not sure about your second remark but we'll see that when implementing deletion on each app.

File: src/app/mork/models/users.py (hunk 1 above)
Comment (wilbrdt) [reply src/app/mork/models/users.py:41]: You're right!

File: src/app/mork/models/users.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+
+class ServiceStatus(Base):
```
Comment (quitterie-lcs) [inline src/app/mork/models/users.py:44]: I would rename the table as it concerns the user deletion status on a specific service. On first reading, when called on the `read_users` method, I thought the table concerns the table specifically.
What do you think of "UserDeletionStatus" or "UserServiceStatus"

File: src/app/mork/models/users.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        Index("idx_service_status", "service_name", "status"),
+    )
+
+    id: Mapped[UUID] = mapped_column(
+        UUID(as_uuid=True), primary_key=True, default=uuid4
+    )
+    user_id: Mapped[UUID] = mapped_column(
+        ForeignKey("users.id", ondelete="CASCADE"), nullable=False
+    )
+    service_name: Mapped[ServiceName] = mapped_column(String(10), nullable=False)
```
Comment (quitterie-lcs) [inline src/app/mork/models/users.py:65]: Similar to `status` typing, you can type with `ServiceName` enum values
```suggestion
    service_name: Mapped[ServiceName] = mapped_column(Enum(ServiceName), nullable=False)

```

File: src/app/mork/models/users.py (hunk 3 above)
Comment (wilbrdt) [reply src/app/mork/models/users.py]: You're right! Choosing `UserServiceStatus`

File: src/app/mork/tests/api/v1/routers/test_users.py [hunk 5]
```diff
@@ -0,0 +1,607 @@
...
+    UserFactory._meta.sqlalchemy_session = db_session
+
+    # Create 200 users that need to be deleted on all services
+    number_users = 200
+    UserFactory.create_batch(number_users)
+
+    response = await http_client.get(
+        "/v1/users",
+        headers=auth_headers,
+        # params={"deletion_status": "to_delete", "service": "ashley"},
```
Comment (quitterie-lcs) [inline src/app/mork/tests/api/v1/routers/test_users.py:47]:
```suggestion
```

File: src/app/mork/tests/api/v1/routers/test_users.py [hunk 6]
```diff
@@ -0,0 +1,607 @@
...
+    assert len(total_requests) == 5
+
+
+@pytest.mark.anyio
+async def test_user_read(db_session, http_client: AsyncClient, auth_headers: dict):
+    """Test the behavior of retrieving one user."""
+    UserServiceStatusFactory._meta.sqlalchemy_session = db_session
+    UserFactory._meta.sqlalchemy_session = db_session
+
+    # Create one user that need to be deleted on all services
```
Comment (quitterie-lcs) [inline src/app/mork/tests/api/v1/routers/test_users.py:252]:
```suggestion
    # Create one user that needs to be deleted on all services
```

File: src/app/mork/tests/api/v1/routers/test_users.py [hunk 7]
```diff
@@ -0,0 +1,607 @@
...
+):
+    """Test the behavior of updating a user status for an invalid user id."""
+    UserServiceStatusFactory._meta.sqlalchemy_session = db_session
+    UserFactory._meta.sqlalchemy_session = db_session
+
+    # Create some users in the database
+    number_users = 10
+    UserFactory.create_batch(number_users)
+
+    # Attempt to read a user with an invalid ID
```
Comment (quitterie-lcs) [inline src/app/mork/tests/api/v1/routers/test_users.py:526]:
```suggestion
    # Attempt to update an user with an invalid ID
```

File: src/app/mork/tests/api/v1/routers/test_users.py [hunk 8]
```diff
@@ -0,0 +1,607 @@
...
+    response = await http_client.patch(
+        f"/v1/users/{invalid_id}/status/ashley",
+        headers=auth_headers,
+        json={"deletion_status": "deleted"},
+    )
+
+    assert response.status_code == 422
+
+    # Attempt to read an user with a nonexistent ID
```
Comment (quitterie-lcs) [inline src/app/mork/tests/api/v1/routers/test_users.py:535]:
```suggestion
    # Attempt to update an user with a nonexistent ID
```

**Review by lunika — APPROVED** [review]

**Review by quitterie-lcs — APPROVED** [review]

**Pull request merged**